README.md
![ForecastIO](header.png)

![Swift](https://img.shields.io/badge/Swift-5.0-brightgreen.svg)
[![CI Status](http://img.shields.io/travis/sxg/ForecastIO.svg?style=flat)](https://travis-ci.org/sxg/ForecastIO)
[![Documentation](https://img.shields.io/cocoapods/metrics/doc-percent/ForecastIO.svg)](http://cocoadocs.org/docsets/ForecastIO/)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
[![Version](https://img.shields.io/cocoapods/v/ForecastIO.svg?style=flat)](http://cocoapods.org/pods/ForecastIO)
[![License](https://img.shields.io/cocoapods/l/ForecastIO.svg?style=flat)](http://cocoapods.org/pods/ForecastIO)
[![Platform](https://img.shields.io/cocoapods/p/ForecastIO.svg?style=flat)](http://cocoapods.org/pods/ForecastIO)

## Requirements

To use ForecastIO, all you need is an API key for the [Dark Sky API](https://darksky.net/dev/). ForecastIO supports iOS (≥9.0), macOS (≥10.10), watchOS (≥2.0), and tvOS (≥9.0).

## Installation

ForecastIO is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your `Podfile`:

```ruby
pod "ForecastIO"
```

To integrate using [Carthage](https://github.com/Carthage/Carthage), specify it in your `Cartfile`:

```ogdl
github "sxg/ForecastIO"
```

### Swift 2.3

A Swift 2.3 compatible version of ForecastIO is available on the [swift2.3](https://github.com/sxg/ForecastIO/tree/swift2.3) branch. This version is frozen at the [ForecastIO 2.1.1 API](http://cocoadocs.org/docsets/ForecastIO/2.1.1/) and no longer supported. To install the Swift 2.3 compatible version of ForecastIO, simply add the following line to your `Podfile`:

```ruby
pod "ForecastIO", :git => "https://github.com/sxg/ForecastIO.git", :branch => "swift2.3"
```

To integrate the Swift 2.3 compatible version using Carthage, specify it in your `Cartfile`:

```ogdl
github "sxg/ForecastIO" "swift2.3"
```

## Documentation & Unit Tests

The full documentation for ForecastIO is available on [CocoaDocs](http://cocoadocs.org/docsets/ForecastIO).

ForecastIO includes a full suite of unit tests with 100% code coverage.

## Usage

First, create a `DarkSkyClient` with your API key:

```swift
import ForecastIO
...
let client = DarkSkyClient(apiKey: "YOUR_API_KEY_HERE")
```

You can choose units that you want `Forecast` responses to use:

```swift
client.units = .si
```

The following units are supported:
- SI
- US (default)
- Canadian
- UK
- Auto (uses the local units for the location for which you are requesting weather data)

More details about the units of each property are available [in the Dark Sky API docs](https://darksky.net/dev/docs/forecast).

You can also choose the language that you want `Forecast` responses to use:

```swift
client.language = .english
```

Many languages are supported (a full list is available [here](https://darksky.net/dev/docs/forecast)). If no language is specified, English is used as the default.

With the `DarkSkyClient`, you can make two kinds of requests. The first will get the current `Forecast` for a particular location:

```swift
let myLoc = CLLocationCoordinate2D(latitude: myLat, longitude: myLon)
client.getForecast(location: myLoc) { result in
    switch result {
    case .success(let currentForecast, let requestMetadata):
        // We got the current forecast!
    case .failure(let error):
        // Uh-oh. We have an error!
    }
}
```

The second kind of request is called a time machine request, and it will get a `Forecast` for a particular location at a particular time:

```swift
let myLoc = CLLocationCoordinate2D(latitude: myLat, longitude: myLon)
client.getForecast(location: myLoc, time: myTime) { result in
    switch result {
    case .success(let forecast, let requestMetadata):
        // We got the forecast!
    case .failure(let error):
        // Uh-oh. We have an error!
    }
}
```

The `Forecast` you receive will have metadata as well as `DataPoint`s and `DataBlock`s associated with it. A `DataPoint` such as the `currently` property on `Forecast` represents various weather phenomena occurring at a specific instant in time. A `DataBlock` such as the `minutely`, `hourly`, and `daily` properties on `Forecast` represent the various weather phenomena occurring over a period of time and are represented by an array of `DataPoint`s.

`DataPoint`s and `DataBlock`s contain a large amount of information, and any of these fields can be excluded from the API response through the `excludeFields` parameter of the `getForecast` methods. `excludeFields` is optional and defaults to an empty array, meaning no data will be excluded from the API response. Alternatively, if you need more data, you can set the `extendHourly` parameter of the `getForecast` method to `true` to make the `hourly` property on `Forecast` return hourly data for a full week instead of 24 hours. `extendHourly` is an optional parameter and defaults to `false`. `extendHourly` is not supported on time machine requests.

For a full list of properties defined on all models, consult the [full documentation on CocoaDocs](http://cocoadocs.org/docsets/ForecastIO).

## Author

Satyam Ghodasara, sghodas@gmail.com, [@\_Satyam\_](https://twitter.com/_satyam_)

## License

ForecastIO is available under the MIT license. See the `LICENSE` file for more info.


CHANGELOG.md
# Change Log

ForecastIO adheres to [Semantic Versioning](http://semver.org/).

## 8.0.0

Released on September 19, 2019.

### Updated
- `darkSkyUnavailable` on `Flag` is now a `String?` instead of a `Bool?`

## 7.0.0

Released on September 4, 2019.

### Added
- `precipitationIntensityError` property on `DataPoint`
- Support for more languages

### Updated
- Swift 5 compatibility
- Replaced custom `Result` with Swift's native implementation
- Replaced `latitude` and `longitude` properties in `DarkSkyClient` and `Forecast` with `location` property of type `CLLocationCoordinate2D`

## 6.0.1

Released on April 8, 2019.

### Updated
- `CodingKeys` for precipitation properties on `DataPoint`
- `nearestStations` on `Flag` is now optional
- All unit test JSON files are now included in all unit test targets

## 6.0.0

Released on March 15, 2019.

### Added
- Most structs now conform to `Decodable`
- `temperatureMax`, `temperatureMaxTime`, `temperatureMin`, `temperatureMinTime`, `apparentTemperatureMax`, `apparentTemperatureMaxTime`, `apparentTemperatureMin`, `apparentTemperatureMinTime`, `windGustTime` on `DataPoint`
- New errors to `ForecastIOError`: `.missingData`, `.missingHeaders`, and `.unexpectedError`
- `nearestStation` on `Flag`

### Updated
- Unit test example JSON data
- Unit tests to conform to new and removed properties and `Decodable`

### Removed
- Custom `init` methods on structs that now conform to `Decodable`
- `ForecastIOError.invalidJSON` error
- `darkSkyStations`, `dataPointStations`, `isdStations`, `lampStations`, `metarStations`, and `metnoLicense` from `Flag`

## 5.1.0

Released on October 2, 2018.

### Added
- Support for Hebrew and Latvian

### Updated
- Fixed `apiRequestsCount` and `responseTime` serialization from headers
- Swift 4.2 compatibility

## 5.0.0

Released on March 31, 2018.

### Added
- `uvIndex`, `uvIndexTime`, and `windGust` properties on `DataPoint`
- Support for Danish, Finnish, Japanese, Georgian, Korean, and Romanian
- `CONTRIBUTING.md`, which describes basic guidelines for contributing to this project

### Updated
- Swift 4.1 compatibility
- Max/min temperatures on `DataPoint` are now high/low

## 4.1.0

Released on March 30, 2018.

### Added
- `time`, `regions`, and `severity` properties on `Alert`
- Support for Bulgarian

### Updated
- `Float` properties are now `Double` on `DataPoint` and `Forecast`

## 4.0.0

Released on February 6, 2017.

### Updated
- `expire` property on `Alert` is now optional

## 3.1.0

Released on December 24, 2016.

### Added
- Support for Catalan, Estonian, and Slovenian

### Updated
- References to the `Array` type now use the `[]` shorthand

## 3.0.3

Released on December 18, 2016.

### Removed
- ForecastIO arm scheme

## 3.0.2

Released on October 17, 2016.

### Updated
- Fixed enum docs 

## 3.0.1

Released on October 8, 2016.

### Removed
- Official iOS 8.x support

## 3.0.0

Released on October 6, 2016.

### Added
- `Result<V>` enum type to encapsulate responses
- `RequestMetadata` to represent data about the consumption of the Dark Sky API
- `ForecastIOError` to represent errors in interacting with the Dark Sky API
- Explicit gzip encoding to all Dark Sky API requests

### Updated
- Lowercased all enum cases
- Moved `ForecastField` enum under `Forecast` as `Field`
- Expanded `precip` on properties to `precipitation`
- `DataBlock.data` is no longer optional
- `Alert.description` is no longer optional
- `Flag.sources` and `Flag.units` are no longer optional
- `Flag.units` is now a `Units` enum type
- API endpoint is now `api.darksky.net` instead of `api.forecast.io`
- `getForecast` methods renamed `completion` to `completionHandler`
- `getForecast` methods' `completionHandler`s now return `Result<Forecast>`
- Documentation formatting
- `APIClient` is now `DarkSkyClient`

### Removed
- `Units.uk2` is gone and its value is now represented by `Units.uk`
- `CustomStringConvertible` protocol conformance from all enums
- `Forecast.offset`
- Constants file and `ForecastIOErrorCode` and `ForecastIOErrorDomain`

## 2.1.1

Released on July 25, 2016.

### Updated
- Fixed typo in `README.md` for the language setting code
- Fixed typos in `Language.swift` docs
- Removed `xcov_report` folder

## 2.1.0

Released on July 24, 2016.

### Added
- Support for multiple languages
- More unit tests for enums
- More platforms for unit testing on Travis CI
- CocoaDocs and Carthage badges to `README.md`
- Codecov.io integration

### Updated
- Unit tests are stricter in checking for a valid URL
- Unit tests now use `@testable` to import ForecastIO
- Details on the optional parameters `extendHourly` and `excludeForecastFields` in `README.md`

## 2.0.0

Released on June 18, 2016.

### Added
- Exclusion of fields from `getForecast` (by [miwand](https://github.com/miwand))
- Extension of `hourly` property to include up to 168 hours of data (by [miwand](https://github.com/miwand))
- `ForecastField` to represent `DataPoint`s and `DataBlock`s on `Forecast` responses (by [miwand](https://github.com/miwand))

### Removed
- Support for OS X 10.9 - deployment target for OS X is now 10.10

### Updated
- Description of `DataPoint`s and `DataBlock`s in `README.md`
- Example project uses `excludeForecastFields` parameter

## 1.6.1

Released on May 12, 2016.

#### Updated
- Fixed issue where the project's interval version number wasn't updated
- Carthage installation section in `README.md`
- Unit test section in `README.md`

## 1.6.0

Released on May 4, 2016.

#### Added
- `ForecastIOErrorDomain` error domain
- `ForecastIOErrorBadJSON` error type

#### Updated
- `APIClient` methods provide more detailed errors
- Unit tests have 100% code coverage
- Travis CI unit tests run on latest versions of iOS, watchOS, and tvOS
- CocoaPod dependencies for unit tests
- Lowered the example project's deployment target to iOS 8.0

## 1.5.0

Released on Jan 24, 2016.

#### Added
- This `CHANGELOG.md` file
- `Package.swift` manifest file for Swift Package Manager

Source/Forecast.swift
//
//  Forecast.swift
//  ForecastIO
//
//  Created by Satyam Ghodasara on 7/18/15.
//  Copyright (c) 2015 Satyam. All rights reserved.
//

import Foundation
import CoreLocation

/// The weather data for a location at a specific time.
public struct Forecast: Decodable {
    
    /// The requested location.
    public let location: CLLocationCoordinate2D
    
    /// The IANA timezone name for the requested location (e.g. "America/New_York"). Rely on local user settings over this property.
    public let timezone: String
    
    /// Severe weather `Alert`s issued by a governmental weather authority for the requested location.
    public let alerts: [Alert]?
    
    /// Metadata for the request.
    public let flags: Flag?
    
    /// The current weather conditions at the requested location.
    public let currently: DataPoint?
    
    /// The minute-by-minute weather conditions at the requested location for the next hour aligned to the nearest minute.
    public let minutely: DataBlock?
    
    /// The hourly weather conditions at the requested location for the next two days aligned to the top of the hour.
    public let hourly: DataBlock?
    
    /// The daily weather conditions at the requested location for the next week aligned to midnight of the day.
    public let daily: DataBlock?
    
    /// Data fields associated with a `Forecast`.
    public enum Field: String, Decodable {
        
        /// Current weather conditions.
        case currently = "currently"
        
        /// Minute-by-minute weather conditions for the next hour.
        case minutely = "minutely"
        
        /// Hour-by-hour weather conditions for the next two days by default but can be exte1nded to one week.
        case hourly = "hourly"
        
        /// Day-by-day weather conditions for the next week.
        case daily = "daily"
        
        /// Severe weather alerts.
        case alerts = "alerts"
        
        /// Miscellaneous metadata.
        case flags = "flags"
    }
    
    /// Creates a new `Forecast` from JSON using `Decodable`.
    ///
    /// - parameter decoder: JSON decoder (provided by `Decodable`).
    ///
    /// - returns: A new `Forecast` configured by the requested JSON.
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.timezone = try container.decode(String.self, forKey: .timezone)
        self.alerts = try container.decodeIfPresent([Alert].self, forKey: .alerts)
        self.flags = try container.decodeIfPresent(Flag.self, forKey: .flags)
        self.currently = try container.decodeIfPresent(DataPoint.self, forKey: .currently)
        self.minutely = try container.decodeIfPresent(DataBlock.self, forKey: .minutely)
        self.hourly = try container.decodeIfPresent(DataBlock.self, forKey: .hourly)
        self.daily = try container.decodeIfPresent(DataBlock.self, forKey: .daily)
        let lat = try container.decode(Double.self, forKey: .latitude)
        let lon = try container.decode(Double.self, forKey: .longitude)
        self.location = CLLocationCoordinate2D(latitude: lat, longitude: lon)
    }
    
    /// Map `Forecast`'s properties to JSON keys.
    private enum CodingKeys: String, CodingKey {
        case timezone, alerts, flags, currently, minutely, hourly, daily
        case latitude, longitude
    }
    
}
